### PR TITLE
Disk controller vmware deploy as is

### PR DESCRIPTION
**Pull request by Spaceman1984**
### Description

This PR...

Fixes: #4344

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

### How Has This Been Tested?

This has been tested by stopping an instance and changing the disk controller setting.
Note - when using a deploy as is template these settings are greyed out.

### CODE REVIEW

**Comment by Spaceman1984** [comment]
@blueorangutan help

**Comment by Spaceman1984** [comment]
@blueorangutan test centos7 vmware-67u3

File: api/src/main/java/org/apache/cloudstack/api/response/UserVmResponse.java [hunk 1]
```diff
@@ -124,6 +124,10 @@
     @Param(description = " an alternate display text of the template for the virtual machine")
     private String templateDisplayText;
 
+    @SerializedName("deployasistemplate")
```
Comment (rhtyd) [inline api/src/main/java/org/apache/cloudstack/api/response/UserVmResponse.java:127]: You want to move this to `ApiConstants....`

Comment (Spaceman1984) [reply api/src/main/java/org/apache/cloudstack/api/response/UserVmResponse.java]: Done

**Review by DaanHoogland — APPROVED** [review]
code looks good, but I cannot judge the view redefinition. extensive testing needed for that

**Comment by DaanHoogland** [comment]
since the user_vm_view changed, do we really want this in 4.15?

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java [hunk 2]
```diff
@@ -2067,9 +2067,7 @@ protected StartAnswer execute(StartCommand cmd) {
                 }
 
                 VirtualMachineDiskInfo matchingExistingDisk = getMatchingExistingDisk(diskInfoBuilder, vol, hyperHost, context);
-                controllerKey = getDiskController(matchingExistingDisk, vol, vmSpec, ideControllerKey, scsiControllerKey);
```
Comment (nvazquez) [inline plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java:2070]: Why is this being removed?

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java [hunk 3]
```diff
@@ -3273,47 +3271,9 @@ private VirtualMachineDiskInfo getMatchingExistingDisk(VirtualMachineDiskInfoBui
         return null;
     }
 
-    private int getDiskController(VirtualMachineDiskInfo matchingExistingDisk, DiskTO vol, VirtualMachineTO vmSpec, int ideControllerKey, int scsiControllerKey) {
```
Comment (nvazquez) [inline plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java:3276]: Same ^^

File: ui/src/components/view/DetailSettings.vue [hunk 4]
```diff
@@ -85,6 +85,7 @@
           <tooltip-button
             :tooltip="$t('label.edit')"
             icon="edit"
+            :disabled="deployasistemplate === true && (item.name === 'rootDiskController' || item.name === 'dataDiskController')"
```
Comment (nvazquez) [inline ui/src/components/view/DetailSettings.vue:88]: I think `deployasistemplate === true` should be enough in this condition

File: ui/src/components/view/DetailSettings.vue [hunk 5]
```diff
@@ -96,6 +97,7 @@
             :title="`${$t('label.delete.setting')}?`"
             @confirm="deleteDetail(index)"
             :okText="$t('label.yes')"
+            :disabled="deployasistemplate === true && (item.name === 'rootDiskController' || item.name === 'dataDiskController')"
```
Comment (nvazquez) [inline ui/src/components/view/DetailSettings.vue:100]: Same as above

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java (hunk 2 above)
Comment (Spaceman1984) [reply plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java:2070]: Unused code

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java (hunk 3 above)
Comment (Spaceman1984) [reply plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java:3276]: Unused code

File: ui/src/components/view/DetailSettings.vue (hunk 4 above)
Comment (Spaceman1984) [reply ui/src/components/view/DetailSettings.vue]: Done

File: ui/src/components/view/DetailSettings.vue (hunk 5 above)
Comment (Spaceman1984) [reply ui/src/components/view/DetailSettings.vue]: Done

**Comment by Spaceman1984** [comment]
@blueorangutan package

**Comment by Spaceman1984** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by rhtyd** [comment]
@nvazquez can you review this PR? thanks.

**Comment by rhtyd** [comment]
Ping @sureshanaparti, @harikrishna-patnala / @nvazquez can you help review this?

File: api/src/main/java/org/apache/cloudstack/api/ApiConstants.java [hunk 6]
```diff
@@ -830,6 +830,7 @@
     public static final String BOOT_MODE = "bootmode";
     public static final String BOOT_INTO_SETUP = "bootintosetup";
     public static final String DEPLOY_AS_IS = "deployasis";
+    public static final String DEPLOY_AS_IS_TEMPLATE = "deployasistemplate";
```
Comment (shwstppr) [inline api/src/main/java/org/apache/cloudstack/api/ApiConstants.java:833]: @Spaceman1984 is this still used?

**Review by shwstppr — CHANGES_REQUESTED** [review]
@Spaceman1984 I think we also need to change updateVirtualMachine API not allow changing details for VM deployed with deployasis template.
It was fine in UI
Screenshot from 2021-05-13 14-59-46
But I was able to update details from API,
```
(localcloud) SBCM5> > update virtualmachine id=d4672813-1c13-4d9d-b91a-6b6ee7717c0c details[0].dataDiskController=pvscsi
{
  "virtualmachine": {
    "account": "admin",
    "affinitygroup": [],
    "cpunumber": 1,
    "cpuspeed": 500,
    "created": "2021-05-13T09:26:22+0000",
    "details": {
      "dataDiskController": "pvscsi"
    },
    "displayname": "t11",
    "displayvm": true,
    "domain": "ROOT",
    "domainid": "89597c05-92b4-11eb-9b07-1e0053000169",
    "guestosid": "89703585-92b4-11eb-9b07-1e0053000169",
    "haenable": false,
    "hypervisor": "VMware",
    "id": "d4672813-1c13-4d9d-b91a-6b6ee7717c0c",
    "instancename": "i-2-37-VM",
....
```
Screenshot from 2021-05-13 14-59-55

File: api/src/main/java/org/apache/cloudstack/api/ApiConstants.java (hunk 6 above)
Comment (Spaceman1984) [reply api/src/main/java/org/apache/cloudstack/api/ApiConstants.java]: No

**Comment by Spaceman1984** [comment]
Changes made @shwstppr

**Comment by Spaceman1984** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 7]
```diff
@@ -2521,6 +2521,13 @@ public UserVm updateVirtualMachine(UpdateVMCmd cmd) throws ResourceUnavailableEx
         String extraConfig = cmd.getExtraConfig();
 
         UserVmVO vmInstance = _vmDao.findById(cmd.getId());
+        if (details.size() > 0) {
```
Comment (nvazquez) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:2524]: Maybe `MapUtils.isNotEmpty(details)` to prevent possible NPE?

Comment (Spaceman1984) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java]: Sure

**Comment by Spaceman1984** [comment]
@blueorangutan package

**Comment by Spaceman1984** [comment]
@blueorangutan package

**Comment by Spaceman1984** [comment]
@blueorangutan test centos7 vmware-67u3

**Review by nvazquez — APPROVED** [review]

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 8]
```diff
@@ -2521,6 +2521,13 @@ public UserVm updateVirtualMachine(UpdateVMCmd cmd) throws ResourceUnavailableEx
         String extraConfig = cmd.getExtraConfig();
 
         UserVmVO vmInstance = _vmDao.findById(cmd.getId());
+        if (MapUtils.isNotEmpty(details)) {
```
Comment (shwstppr) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:2524]: @Spaceman1984 should we also check cleanupDetails param here?

Comment (Spaceman1984) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java]: Yes, that would be good.

**Comment by Spaceman1984** [comment]
@blueorangutan package

**Review by shwstppr — APPROVED** [review]
LGTM. Verified both UI and API
Screenshot from 2021-05-17 14-02-10


```
(localcloud) SBCM5> > list templates id=888f0a52-e34b-4193-8cfb-4b7d3ce42753 templatefilter=self filter=id,name,deployasis
{
  "count": 1,
  "template": [
    {
      "deployasis": true,
      "id": "888f0a52-e34b-4193-8cfb-4b7d3ce42753",
      "name": "mac"
    }
  ]
}
(localcloud) SBCM5> > list virtualmachines id=bb535507-ffb3-48dc-85ed-6aafd6c18512 filter=id,name,details,templateid,templatename,
{
  "count": 1,
  "virtualmachine": [
    {
      "details": {
        "Message.ReservedCapacityFreed.Flag": "false",
        "cpuOvercommitRatio": "2.0",
        "dataDiskController": "osdefault",
        "memoryOvercommitRatio": "1.0",
        "rootDiskController": "lsilogic"
      },
      "id": "bb535507-ffb3-48dc-85ed-6aafd6c18512",
      "name": "t2",
      "templateid": "888f0a52-e34b-4193-8cfb-4b7d3ce42753",
      "templatename": "mac"
    }
  ]
}
(localcloud) SBCM5> > update virtualmachine id=bb535507-ffb3-48dc-85ed-6aafd6c18512 cleanupdetails=true 
🙈 Error: (HTTP 530, error code 4250) Detail settings are read from OVA, it cannot be changed by API call.
(localcloud) SBCM5> > update virtualmachine id=bb535507-ffb3-48dc-85ed-6aafd6c18512 details[0].dataDiskController=pvscsi
🙈 Error: (HTTP 530, error code 4250) Detail settings are read from OVA, it cannot be changed by API call.
```

**Comment by Spaceman1984** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by Spaceman1984** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by rhtyd** [comment]
Let's merge this after a smoketest run kicked above ^^ and if @vladimirpetrov agrees, thanks.

**Comment by Spaceman1984** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by Spaceman1984** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by rhtyd** [comment]
@vladimirpetrov are you lgtm on it?

**Review by vladimirpetrov — APPROVED** [review]
Ok, the description is not very detailed so let me improve that. There are actually two fixes here:
1. Passing the disk controller setting to vSphere - fix confirmed, tested with root and data disks.
2. When using a deploy-as-is template, all VM settings should be disabled - fix confirmed.

LGTM

**Pull request merged**